draft-ietf-ccwg-bbr.md

```diff
@@ -1168,50 +1174,29 @@ After each packet transmission, the sender executes the following steps:
 
 #### Upon receiving an ACK {#upon-receiving-an-ack}
 
-When an ACK arrives, the sender invokes GenerateRateSample() to fill in a
-rate sample. For each  packet that was newly acknowledged, UpdateRateSample()
-updates the rate sample based on a snapshot of connection delivery information
-from the time at which the packet was last transmitted. UpdateRateSample()
-is invoked multiple times when a stretched ACK acknowledges multiple data
-packets. In this case we use the information from the most recently sent
-packet, i.e., the packet with the highest "P.delivered" value.
+When an ACK arrives, the connection first calls InitRateSample() to initialize
+the per-ACK RateSample RS:
 
 ~~~~
-  /* Upon receiving ACK, fill in delivery rate sample RS. */
-  GenerateRateSample(RS):
-    for each newly acknowledged packet P
-      UpdateRateSample(P, RS)
-
-    /* Clear app-limited field if bubble is ACKed and gone. */
-    if (C.app_limited and C.delivered > C.app_limited)
-      C.app_limited = 0
-
-    if (RS.prior_time == 0)
-      return false  /* nothing delivered on this ACK */
-
-    /* Use the longer of the send_elapsed and ack_elapsed */
-    RS.interval = max(RS.send_elapsed, RS.ack_elapsed)
-
-    RS.delivered = C.delivered - RS.prior_delivered
-
-    /* Normally we expect interval >= MinRTT.
-     * Note that rate may still be overestimated when a spuriously
-     * retransmitted skb was first (s)acked because "interval"
-     * is under-estimated (up to an RTT). However, continuously
-     * measuring the delivery rate during loss recovery is crucial
-     * for connections that suffer heavy or prolonged losses.
-     */
-    if (RS.interval <  C.min_rtt)
-      RS.interval = -1
-      return false  /* no reliable sample */
-
-    if (RS.interval != 0)
-      RS.delivery_rate = RS.delivered / RS.interval
+  /* Initialize the rate sample generated using the ACK being processed. */
+  InitRateSample():
+    RS.rtt           = -1
+    RS.has_data      = false
+    RS.prior_time    = 0
+    RS.interval      = 0
+    RS.delivery_rate = 0
+~~~~
 
-    return true;  /* we filled in rs with a rate sample */
+Next, for each packet that was newly acknowledged, the connection calls
+UpdateRateSample() to update the per-ACK rate sample based on a snapshot of
+connection delivery information from the time at which the packet was last
```

Comment:
Last transmitted is a bit odd for non-TCP transports, since in QUIC a packet is only transmitted once.  Can we just drop 'last' and leave it as an implementation detail?

```suggestion
connection delivery information from the time at which the packet was
```